dialogue-core/src/main/java/com/palantir/dialogue/core/LeakDetectingChannel.java

```diff
@@ -0,0 +1,187 @@
+        }
+
+        void disarm() {
+            armed = false;
+        }
+
+        @Override
+        @SuppressWarnings("NoFinalizer")
+        protected void finalize() throws Throwable {
+            if (armed) {
+                metrics.responseLeak()
+                        .channelName(channelName)
+                        .serviceName(endpoint.serviceName())
+                        .endpoint(endpoint.endpointName())
+                        .build()
+                        .mark();
+                if (creationTrace == null) {
+                    log.warn(
+                            "Detected a leaked response from service {} endpoint {} on channel {}. Enable trace "
+                                    + "logging to record stack traces.",
+                            SafeArg.of("service", endpoint.serviceName()),
+                            SafeArg.of("endpoint", endpoint.endpointName()),
+                            SafeArg.of("channel", channelName));
+                } else {
+                    log.warn(
+                            "Detected a leaked response from service {} endpoint {} on channel {}",
+                            SafeArg.of("service", endpoint.serviceName()),
+                            SafeArg.of("endpoint", endpoint.endpointName()),
+                            SafeArg.of("channel", channelName),
+                            creationTrace);
+                }
+                response.close();
+            }
+            super.finalize();
+        }
+
+        @Override
+        public String toString() {
+            return "LeakDetector{endpoint=" + endpoint + ", response=" + response + ", armed=" + armed + '}';
+        }
+    }
+
+    private static final class LeakDetectingInputStream extends FilterInputStream {
+
+        private final LeakDetector leakDetector;
+
+        LeakDetectingInputStream(InputStream delegate, LeakDetector leakDetector) {
+            super(delegate);
+            this.leakDetector = leakDetector;
+        }
+
+        @Override
+        public void close() throws IOException {
+            leakDetector.disarm();
+            super.close();
+        }
+
+        @Override
+        public String toString() {
+            return "LeakDetectingInputStream{leakDetector=" + leakDetector + ", in=" + in + '}';
+        }
+    }
+
+    private static final class LeakDetectingResponse implements Response {
+
+        private final Response delegate;
+        private final LeakDetector leakDetector;
+
+        @Nullable
+        private InputStream leakDetectingStream;
+
+        LeakDetectingResponse(Response delegate, LeakDetector leakDetector) {
+            this.delegate = delegate;
+            this.leakDetector = leakDetector;
+        }
+
+        @Override
+        public InputStream body() {
+            if (leakDetectingStream == null) {
+                leakDetectingStream = new LeakDetectingInputStream(delegate.body(), leakDetector);
```

Comment:
why the lazy initialization here?

Reply:
In the case delegate.body throws, I figure this will produce a better stack trace than failing when the response is wrapped.
It’s probably fine either way though.